src/hotspot/share/opto/addnode.cpp

```diff
@@ -769,6 +771,553 @@ uint AddPNode::match_edge(uint idx) const {
   return idx > Base;
 }
 
+/* MergeLoads optimization
+ * Optimize multiple loads into a merged load, like below java code:
+ *
+ *      return ((array[offset    ] & 0xff) << 24)
+ *           | ((array[offset + 1] & 0xff) << 16)
+ *           | ((array[offset + 2] & 0xff) <<  8)
+ *           | ((array[offset + 3] & 0xff)      );
+ *
+ * The C2 IR graph is like:
+ *
+ *       Mem
+ *        |---------+---------+----------+
+ *        |         |         |          |
+ *      LoadB1    LoadB2    LoadB3     LoadB4
+ *        |         |         |          |
+ *        |       LShift    LShift     LShift
+ *        |         |         |          |
+ *        +-> Or1 <-+         |          |
+ *             |              |          |
+ *             +---> Or2 <----+          |
+ *                    |                  |
+ *                    +-----> Or3 <------+
+ *
+ * It is transformed as a merged LoadI, which replace the Or3 node.
+ *
+ * Because array value is logical and with constant 0xff/0xffff, LoadS/LoadB is converted to an unsigned load
+ * and 'And' node is eliminated in previous IGVN phase. Please check AndINode::Ideal for reference
+ *
+ */
+class MergePrimitiveLoads;
+
+/*
+ * A LoadNode and a combine Node pair which represent an item for merging,
+ * now only OrI/OrL are supported as combine operator, in future we can extend to add/xor
+ *
+ * Note: combine operator can be shared in different MergeLoadInfo
+ */
+class MergeLoadInfo : ResourceObj {
+  friend MergePrimitiveLoads;
+private:
+  LoadNode* const    _load;
+  Node*     const _combine;
+  int       const   _shift;
+public:
+  MergeLoadInfo(LoadNode* load, Node* combine, int shift) : _load(load), _combine(combine), _shift(shift) {}
+
+#ifdef ASSERT
+  void dump() {
+    tty->print_cr("MergeLoadInfo: load: %d, combine: %d, shift: %d",
+                  _load->_idx, _combine->_idx, _shift);
+  }
+#endif
+};
+
+typedef GrowableArray<MergeLoadInfo*> MergeLoadInfoList;
```

Comment:
Do you need to have this Resource allocated? Why not just have the elements in-place? And then make the `MergeLoadInfo` a `StackObj`. That would remove the extra pointer indirection.

Reply:
I choose `GrowableArray` for convenience. And it can be changed as a stack allocate data.

Reply:
Ah, we have a misunderstanding. I was asking why not 
```suggestion
typedef GrowableArray<MergeLoadInfo> MergeLoadInfoList;
```
i.e. allocate the elements of the array directly in the array (in-place), rather than allocating separate elements.

Reply:
I changed as `GrowableArray<MergeLoadInfo>` .